### PR TITLE
PR from dev to master, dev branch to be removed

### LINKED ISSUE #39: Personal and "dev" branches
What is the idea behind your personal branches as well the `dev` one? This practice is generally considered a red flag :red_circle: .

From what I have seen, you are merging to `dev` without any code reviews (e.g #37, #32) and at some point you do large(r) merges to `master` which naturally cannot be code-reviewed as effectively.

If you are doing this to get around the need of having to code review, I will add compulsory code reviews for the `dev` branch too. :scream: 
¯\_(ツ)_/¯

### PR DESCRIPTION
**Pull request by leithhobson**
This PR is to bring master up to the current state of dev to resolve #39

### CODE REVIEW

**Comment by platisd** [comment]
:tada:

File: Car/TeamNull SmartCar/src/util/HardwareUtil.cpp [hunk 1]
```diff
@@ -77,11 +83,16 @@ void initialiseSensors() {
...
-    while (1) {
-    }
+    blink(2);
+    pinMode(frontSensorPin, OUTPUT);
+    digitalWrite(frontSensorPin, LOW);
+
+    delay(1500);
+    digitalWrite(frontSensorPin, HIGH);
+    delay(200);
+    pinMode(frontSensorPin, INPUT);
```
Comment (platisd) [inline Car/TeamNull SmartCar/src/util/HardwareUtil.cpp]: Why are you flipping the pins and not using the Smartcar library?

**Review by platisd — CHANGES_REQUESTED** [review]

File: Car/TeamNull SmartCar/src/util/HardwareUtil.cpp (hunk 1 above)
Comment (leithhobson) [reply Car/TeamNull SmartCar/src/util/HardwareUtil.cpp]: Perhaps ignorance. I am not aware that the library supports the VL53L0X sensors directly, and further did not find a section in the documentation referring to the setting of pins? So I'm not entirely sure how we should be using the library in this case? Could you please point us to a section in the documentation, or otherwise guide us, so that this can be resolved?

File: Car/TeamNull SmartCar/src/util/HardwareUtil.cpp [hunk 2]
```diff
@@ -61,12 +61,18 @@ void initialiseSensors() {
...
```
Comment (platisd) [inline Car/TeamNull SmartCar/src/util/HardwareUtil.cpp]: What does this do?

File: Car/TeamNull SmartCar/src/util/HardwareUtil.cpp (hunk 1 above)
Comment (platisd) [inline Car/TeamNull SmartCar/src/util/HardwareUtil.cpp]: What does this (duplicated code) do?

Comment (platisd) [reply Car/TeamNull SmartCar/src/util/HardwareUtil.cpp]: At first I thought you were trying to use an ultrasonic sensor but that's not the case. I have no idea what you are trying to do with your code. Left some additional comments to be more exact.

**Review by platisd — DISMISSED** [review]

**Review by platisd — COMMENTED** [review]
```cpp
int frontSensorPin = 33;
int rearSensorPin = 32;
```

What is connected on those pins?! I really hope it's not the VL53L0X sensors because I am not sure how tolerant they are of 5V. :cry:

File: Car/TeamNull SmartCar/src/util/HardwareUtil.cpp (hunk 2 above)
Comment (leithhobson) [reply Car/TeamNull SmartCar/src/util/HardwareUtil.cpp]: Basically, when the car is powered by USB, everything boots correctly, however, when booting from batteries, the VL53L0Xs fail to initialise, this repeated powering them up and down was intended to see if they would eventually power-up successfully, after repeated attempts.
We believe therefore that the battery-based power supply has some sort of an issue where the ground is not at 0v, or something along those lines, although it is not easy to test this theory, as we do not have access to an oscilloscope, and the multimeter that we have does not seem to provide any value.

Configuring the pin as an output, with a low state will not damage the sensor, and, as the sensor has an internal pull-up resistor, configuring the pin as INPUT will cause it to be pulled up to 3.3v (to answer partially your next question)

The code configuring the pin high was never run on the car, as we had intended first to research if the board would handle this, before running it.

File: Car/TeamNull SmartCar/src/util/HardwareUtil.cpp (hunk 1 above)
Comment (leithhobson) [reply Car/TeamNull SmartCar/src/util/HardwareUtil.cpp]: Yes, it's duplicated, will not be once we pass the debugging stage of trying to get two VL53L0Xs to work simultaneously.

**Comment by leithhobson** [comment]
> ```c++
> int frontSensorPin = 33;
> int rearSensorPin = 32;
> ```
> 
> What is connected on those pins?! I really hope it's not the VL53L0X sensors because I am not sure how tolerant they are of 5V. cry

Yes, they are connected to the VL53L0X sensors, although, as mentioned before, the sensors have not been driven to 5v.
They are specifically connected to the XSHUT pins on the respective sensors

File: Car/TeamNull SmartCar/src/util/HardwareUtil.cpp (hunk 2 above)
Comment (platisd) [reply Car/TeamNull SmartCar/src/util/HardwareUtil.cpp]: >  however, when booting from batteries, the VL53L0Xs fail to initialise

You are the first group that comes up with this problem and generally considering the circumstances I find it highly unlikely there is an actual problem.
How are you connected to the sensors? Have you placed the sensors in the pre-determined sockets?

> Configuring the pin as an output, with a low state will not damage the sensor, and, as the sensor has an internal pull-up resistor, configuring the pin as INPUT will cause it to be pulled up to 3.3v (to answer partially your next question)

Yet again, how have you connected the sensors? Is the ESP32 board attached to the car? If so, then the pins that are exposed with the female sockets use 5V, NOT 3.3V. 

Please clarify or show your connections.

Comment (leithhobson) [reply Car/TeamNull SmartCar/src/util/HardwareUtil.cpp]: (Please excuse any typo's, working on a laptop that doesn't have English installed)
- One VL53L0X works perfectly, as expected
- However, in order to use two (or more), as per the sensor's documentation, you have to cycle the XSHUT pins sequentially, in order to give each sensor a unique address. - I am not aware of any other groups that are currently using both provided VL53L0X sensors.

- The sensors are in the designated sockets
- However, according to the schematics of the circuit board that are provided, the XSHUT pins on the headers do not appear to be traced to any pins on the ESP32, and so we have connected them to the exposed female pin headers. As I said, the code as is, driving the pins high has not been run on the car, because of the 5V issue.
The pins are raised to 3.3V by configuring them as inputs, in combination with the sensors internal pull-up resistors.

- TL/DR: Sensor is in designated header, with additional wire from XSHUT to a female pin header port, not being driven high. One sensor works, multiple do not.

Not sure if this clarifies the connection, alternatively, I can get a picture of the connection.

Comment (platisd) [reply Car/TeamNull SmartCar/src/util/HardwareUtil.cpp]: Oh yes it does and I just discovered that those sensors have a really **stupid** design. I was under the assumption that once you set an address the address persists which is the behavior of other i2c sensors. 😩😩
I guess I'm also stupid for not trying this out earlier assuming it'd work. 😓😓😓
I mean, I selected those sensors literally because I thought I could put all of them on the i2c bus and waste no pins. Now it appears we'd have to waste one pin for each sensor if multiple were used.

In any case, unless you use the 3.3V pins on the board (which would be pretty hard considering they don't have female sockets) please don't put high voltage on the xshut pin unless you read somewhere it's OK to do so. If you need multiple sensors, use the ultrasonics.

Comment (leithhobson) [reply Car/TeamNull SmartCar/src/util/HardwareUtil.cpp]: Glad we're on the same page now @platisd 
Yes, rather silly that the adresses are not persistent haha, not quite sure who came up with that, as it means you always need to waste a pin per sensor, defeating the i2c pin efficiency.

- Would you be willing to indicate on which pin of the level converters we may solder? Or alternatively, would you be willing to solder this for us, if we bring the car to you?

- We intend using the sonar sensors to detect obstacles to the right, and left of straight ahead, so that the SmartCar might be able to actively avoid objects, and so we would like to use both LiDAR sensors, and not fall back to the sonar ones, if possible.

Comment (platisd) [reply Car/TeamNull SmartCar/src/util/HardwareUtil.cpp]: Let's not go into solutions that involve soldering.

> This pin is an active-low shutdown input; the board pulls it up to VDD to enable the sensor by default. Driving this pin low puts the sensor into hardware standby. This input is not level-shifted.
> [Source](https://www.pololu.com/product/2490)

Alright, so it looks like we can get away with the following:
* To put the sensor in standby mode the pin should be set as an `OUTPUT` and then `LOW` 
* To have the sensor working, the pin should be set as an `INPUT`

In other words, the pin is never set `HIGH` from the microcontroller's side and therefore no 5V should ever reach the sensor.

Comment (leithhobson) [reply Car/TeamNull SmartCar/src/util/HardwareUtil.cpp]: Yes, this is exactly how it was done, as I mentioned, the code as in this pull request was never loaded onto the car, as we were still investigating the possibility of safely using 5V.

When the car is powered by USB, the sensors initialise perfectly, first time, however, when powered by the batteries on the car, the sensors are unable to be initialised, which we suspect has some strange reference voltage issue.
The multimeter that we have, has however not provided useful insight into this, as it's own presence is enough to cause the sensors to initialise correctly, when trying to measure voltages.

Comment (platisd) [reply Car/TeamNull SmartCar/src/util/HardwareUtil.cpp]: I'll try it out today or tomorrow on my setup and will let you know if I find something interesting.

Comment (leithhobson) [reply Car/TeamNull SmartCar/src/util/HardwareUtil.cpp]: > 
> 
> I'll try it out today or tomorrow on my setup and will let you know if I find something interesting.

Thank you, looking forward to the result

Comment (platisd) [reply Car/TeamNull SmartCar/src/util/HardwareUtil.cpp]: Just to verify, did you try the car with batteries on and a *single* sensor?

Comment (leithhobson) [reply Car/TeamNull SmartCar/src/util/HardwareUtil.cpp]: Yes, on batteries, with a single sensor, the car is able to initialse correctly, my belief (unproven) is that the ESP32 pin set as output, low, is not going low enough to reset the sensor, when powered by batteries. If this is the case, it would hold that one sensor will initialise correctly, but two will not.

Comment (leithhobson) [reply Car/TeamNull SmartCar/src/util/HardwareUtil.cpp]: Hi @platisd, any luck with your sensors? :)
TeamNull

Comment (platisd) [reply Car/TeamNull SmartCar/src/util/HardwareUtil.cpp]: Didn't try yet :hear_no_evil:
I'll try this weekend.

**Review by diaistanbuly — APPROVED** [review]
Despite the issues I approved the merge because we need to close the dev branch

File: Car/TeamNull SmartCar/src/util/HardwareUtil.cpp (hunk 2 above)
Comment (leithhobson) [reply Car/TeamNull SmartCar/src/util/HardwareUtil.cpp:73]: @platisd, I am marking this conversation as resolved, so that we can go ahead with the merge, in order to dissolve the DEV branch, but am aware that this issue is still being resolved, although this merge is not the issue as such.

**Comment by leithhobson** [comment]
@platisd, could you please re-review this PR, and approve it, so that we can dispose of the DEV branch. The issue with the sensors are not believed to be code related, and further, the code will no longer ever push the pins high

**Comment by platisd** [comment]
Done

**Comment by leithhobson** [comment]
> Done

Thank you!
